lnwallet/chanfunding/coin_select.go

```diff
@@ -251,3 +254,81 @@ func CoinSelectSubtractFees(feeRate chainfee.SatPerKWeight, amt,
 
 	return selectedUtxos, outputAmt, changeAmt, nil
 }
+
+// CoinSelectUpToAmount attempts to select coins such that we'll select up to
+// maxAmount exclusive of fees if sufficient funds are available. If
+// insufficient funds are available this method selects all available coins.
+func CoinSelectUpToAmount(feeRate chainfee.SatPerKWeight, maxAmount,
+	dustLimit btcutil.Amount, coins []Coin) ([]Coin, btcutil.Amount,
+	btcutil.Amount, error) {
+
+	var (
+		// Track if our coin selection was unsuccessful and whether
+		// we have to start a new round of selecting coins considering
+		// fees.
+		selectSubtractFee = false
+		outputAmount      = maxAmount
+	)
+
+	// First we try to select coins to create an output of the specified
+	// maxAmount with or without a change output that covers the miner fee.
+	selected, changeAmt, err := CoinSelect(
+		feeRate, maxAmount, dustLimit, coins,
+	)
+
+	var errInsufficientFunds *ErrInsufficientFunds
+	if errors.As(err, &errInsufficientFunds) {
+		// If the initial coin selection fails due to insufficient funds
+		// we select our total available balance minus fees.
+		selectSubtractFee = true
+	} else if err != nil {
+		return nil, 0, 0, err
+	}
+
+	// If we determined that our local balance is insufficient we check our
+	// total balance minus fees.
+	if selectSubtractFee {
+		// Get balance from coins.
+		var totalBalance btcutil.Amount
+		for _, coin := range coins {
+			totalBalance += btcutil.Amount(coin.Value)
+		}
+
+		selected, outputAmount, changeAmt, err = CoinSelectSubtractFees(
+			feeRate, totalBalance, dustLimit, coins,
+		)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+	}
+
+	// Make sure that the output amount is below the specified max amount.
+	if outputAmount > maxAmount+dustLimit {
+		return nil, 0, 0, fmt.Errorf("coin selection amount %v "+
+			"exceeded max amount %v", outputAmount, maxAmount)
+	}
+
+	// Attribute the selected extra amount of coins to miner fees if it is
+	// less than the dust limit.
+	if outputAmount > maxAmount {
+		outputAmount = maxAmount
+	}
+
+	// Sanity check the resulting output values to make sure we don't burn a
+	// great part to fees.
+	totalOut := outputAmount + changeAmt
+	sum := func(coins []Coin) btcutil.Amount {
+		sum := btcutil.Amount(0)
+		for _, coin := range coins {
+			sum += btcutil.Amount(coin.Value)
+		}
+
+		return sum
+	}
+	err = sanityCheckFee(totalOut, sum(selected)-totalOut)
```

Comment:
question: In the CoinSelection functions above, there is always a fee sanity check included, I have difficulties imagine a scenario where could fail this sanity checks and why we have to do them here again ?

Reply:
Are you saying that you can't think of a scenario in the coin selection functions that could result in a large fee? Afaiu these checks are there to defend against hidden bugs in the coin selection functions to ensure that a big portion of user funds aren't burned by miner fees.

Reply:
what I meant is, that in both functions prior to this call `CoinSelect` and `CoinSelectSubtractFees` we already make a sanity check, so I am not really sure whether we need this here again

Reply:
Yes agree, however we maybe want to protect against bugs in the interplay between `CoinSelect` and `CoinSelectSubtractFees`? Otherwise we could remove it.

Reply:
IMO you can remove this check, either one of those functions above will do the check, so we are good here, for me doing this check makes things more confusing than bringing more error proneness ?

Reply:
I think we can keep it in case one of the other function changes. I don't find it confusing, it just performs a duplicate check